pkg/api/validate/dynamic/dynamic.go

```diff
@@ -607,34 +612,80 @@ func (dv *dynamic) ValidateSubnets(ctx context.Context, oc *api.OpenShiftCluster
 			)
 		}
 
-		if oc.Properties.ProvisioningState == api.ProvisioningStateCreating {
-			if ss.SubnetPropertiesFormat != nil &&
-				ss.SubnetPropertiesFormat.NetworkSecurityGroup != nil {
-				expectedNsgID, err := subnet.NetworkSecurityGroupID(oc, s.ID)
+		subnetByID[s.ID] = ss
+	}
+	return subnetByID, nil
+}
+
+// checkByoNSG checks whether all the subnets have or don't have NSG attached.
+// when the BYONsg feature flag is on and only some of the subnets are attached with an NSG,
+// it returns an error.  If none of the subnets is attached, it's no longer BYO NSG and the
+// cluster installation process should fall back to using the managed nsg.
+func (dv *dynamic) checkByoNSG(subnetByID map[string]*mgmtnetwork.Subnet) (api.PreconfiguredNSG, error) {
+	noNSGAttached := 0
+	allSubnetsAreAttached := len(subnetByID)
+
+	var attached int
+	for _, subnet := range subnetByID {
+		if subnetHasNSGAttached(subnet) {
+			attached++
+		}
+	}
+	if attached > noNSGAttached && attached < allSubnetsAreAttached {
+		dv.log.Info("BYO NSG: not all subnets are attached")
+		return api.PreconfiguredNSGDisabled,
+			&api.CloudError{
+				StatusCode: http.StatusBadRequest,
+				CloudErrorBody: &api.CloudErrorBody{
+					Code:    api.CloudErrorCodeInvalidLinkedVNet,
+					Message: errMsgNSGNotProperlyAttached,
+				},
+			}
+	}
+	if attached == allSubnetsAreAttached {
+		dv.log.Info("all subnets are attached, BYO NSG")
+		return api.PreconfiguredNSGEnabled, nil // correct setup by customer
+	}
+	dv.log.Info("no subnets are attached, no longer BYO NSG. Fall back to using cluster NSG.")
+	return api.PreconfiguredNSGDisabled, nil
```

Comment:
Let's clean this up just a bit
```suggestion
	// all subnets have an attached NSG
	if attached == len(subnetByID) {
		dv.log.Info("all subnets are attached, BYO NSG")
		return api.PreconfiguredNSGEnabled, nil // correct setup by customer
	}

	// no subnets have attached NSG, fallback
	if attached == 0 {
		dv.log.Info("no subnets are attached, no longer BYO NSG. Fall back to using cluster NSG.")
		return api.PreconfiguredNSGDisabled, nil
	}

	// some subnets have NSGs attached, error out
	return api.PreconfiguredNSGDisabled,
		&api.CloudError{
			StatusCode: http.StatusBadRequest,
			CloudErrorBody: &api.CloudErrorBody{
				Code:    api.CloudErrorCodeInvalidLinkedVNet,
				Message: errMsgNSGNotProperlyAttached,
			},
		}